src/fsharp/Optimizer.fs

```diff
@@ -2543,6 +2543,37 @@ and TryDevirtualizeApplication cenv env (f,tyargs,args,m) =
                 MightMakeCriticalTailcall = false;
                 Info=UnknownValue})
 
+    // Analyze the name of the given symbol and rewrite AST to constant string expression with the name
+    | Expr.Val(vref,_,_),_,_ when valRefEq cenv.g vref cenv.g.nameof_vref ->
+        PostTypeCheckSemanticChecks.tryExtractNameOf args
+        |> Option.map (fun name ->
+            Expr.Const(Const.String name, m, cenv.g.string_ty),
+                { TotalSize = 1
+                  FunctionSize = 1
+                  HasEffect = false
+                  MightMakeCriticalTailcall = false
+                  Info = UnknownValue })
+    // Analyze the name of the given type and rewrite AST to constant string expression with the name
+    | Expr.Val(vref,_,_),_,_ when valRefEq cenv.g vref cenv.g.typenameof_vref ->
+        match tyargs with
+        | (typeName:TType):: _ ->
+            let name =
+                match typeName with 
+                | TType_forall (_tps,ty) -> ty.ToString()
+                | TType_app (tcref, _) -> tcref.CompiledRepresentationForNamedType.FullName
+                | TType_tuple tinst -> "(" + String.concat "," (List.map string tinst) + ")"
+                | TType_fun (d,r) -> "(" + string d + " -> " + string r + ")"
```

Comment:
Using ``string`` here is not correct - please factor this ``convert to a string`` out.

Indeed I think it is best just to fail on tuple types.  For function types, return ``Microsoft.FSharp.Core.FSharpFunc`2``